Liberty Fix WML unit not found for role

GregoryLundberg opened the pull request:
In 1.13.4+dev I noted WML errors in S02 through S04 when attempting to locate a unit for use as an advisor.

I moved the macros over from AOI for locating, promoting and recalling an advisor. For this campaign, the unit types dynamically change; so I added a paramter which specifies the unit type to recruit if no suitable unit could be found. For this, I chose Thug and Thug_Peasant.

As with AOI, these macros select the eldest unit from the board or, of none exist there, from the recall list. This is a slight change from the original: my change allows the Footpad_Peasant to be an advisor, the original omitted this unit as a possibility. I did this so I could simply select a unit which was not one of the two heros and not worry about the specific unit types.

I also changed "Advisor" to "advisor" to align with AOI.

In S03 a [message] used "speaker=Advisor" and never appeared as a result. I changes this to "role=advisor" to have the message appear now that there is reliably an advisor available to deliver the message.

Finally, while testing, I ran into a but in 1.13.4+dev having to do with string-to-integer conversion. This affected S08, only. I changed the word 'FOREVER' to specify '99' turns, instead, to work around this bug. Since the scenario should end on turn 48 (or earlier), that is effectively the same. I can add a revert patch to this PR if a fix appears in the master; but it should not matter if this is never reverted.

GregoryLundberg commented:
With the changes in PR #681 correcting the [option] tags, and these changes, I play-tested Liberty looking for errors or warnings on the console. I ignored a lot of warnings about the UI missing events on mouse clicks. The following are my notes of those issues which seemed to warrant attention:

Liberty, scenario 01, "The Raid"

    warning ai/config: side 2: aspect with id=[version] lacks default config facet!

Liberty, scenario 07, "The Hunters"

    warning ai/config: side 2: aspect with id=[version] lacks default config facet!

    warning ai/config: side 3: aspect with id=[version] lacks default config facet!

During start-up, generally on loading a saved turn or scenario.

    error ai/engine/cpp: side 2 : UNKNOWN aspect[version*composite_aspect]

    error ai/engine/cpp: side 3 : UNKNOWN aspect[version*composite_aspect]

During battle. Note that side 3 has no units remaining and can only pass its turn.



Liberty, scenario 03, "A Strategy Of Hope"
Liberty, scenario 04, "Unlawful Orders"
Liberty, scenario 06, "The Grey Woods"
Liberty, scenario 07, "Glory"

    warning engine: vconfig::get_parsed_config() infinite recursion detected, aborting

Seems to be related to giving the allied AI instructions. The message occurs when presenting the selections which instruct the AI, not after actually selecting an instruction, and not on options which do not effect the AI.

This is sourced from ~/data/core/macros/ai_controller.cfg line 106.

Called from S03 line 30: {AI_CONTROLLER () 1 2 ()}
Called from S04 line 40: {AI_CONTROLLER () 1 2 ()}
Called from S06 line 48: {AI_CONTROLLER () 1 2 ()}
Called from S08 line 52: {AI_CONTROLLER () 1 5,6 ()}



Liberty, scenario 07, "The Hunters"
    error ai/actions: Return value of AI ACTION was not checked. This may cause bugs!

During battle?

Liberty, scenario 8, "Glory"

    error scripting/lua: The &image=col1=col2 syntax is deprecated, use new DescriptionWML instead.

While setting AI for Sir Gwydion (side 6), immediately upon his arrival on turn 4. Note the error did NOT occur when setting the AI for Helicrom (side 5) during turn 1.

    error ai/actions: Error #2007 (E_NOT_REACHED_DESTINATION) in full move by side 2 from location 28,17 to location 27,18

During battle.

gfgtdf commented:
I wonder whether when i foudn a advisor in the recall list with IF_HAVE_UNIT_SET_ROLE  dont we still need to recall them before we can use them in a message?

GregoryLundberg commented:
The advisor only has one speaking line on victory. If there are NO units on the board, but there are units in the recall list, a unit on the recall list will be promoted. I chose not to recall that advisor for this edge condition because the natural flow of the conversation is not adversely affected. This, of course, is in English. If a translation would be effected (I judge that unlikely, but only the translators know), then, yes, a recall should occur.

Note that the translation issue also effects S03. But, if S02 failed to promote, S03 would, as well. Here, the only option would be to recruit a unit.

I originally had a LOT of logic (lifted from AOI) which recalled and recruited, if needed. But I felt I was being too heavy-handed and had to choose a unit to recruit if none existed. I decided to step back and only ensure the spoken line occurred and the free recall of one unit for S03 and S04 actually happened. My intention was to do only the minimum required to implement the clear intention of the previous authors.

Vultraz commented:
The design of previous authors isn't exactly sacrosanct. If you feel there's room for improvement, please by all means improve. :)

GregoryLundberg commented:
How about I get more mileage on this project, first!

I did make a change in my AOI PR, it just seemed so wrong to have the mage/leader recruiting the same elves the the player/elf/leader.

At this point, I'm just trying to run through all the low-level mainline campaigns, fix what I can, and make not of things which might need more work (cf my notes, above).

Vultraz commented on data/campaigns/Liberty/utils/utils.cfg:
```diff
@@ -165,3 +165,27 @@
     [/event]
 #enddef
+
+#define IF_HAVE_UNIT_SET_ROLE sideNumber roleName unitFilter
```
You can simplify this logic like so:

```
    [if]
        [not]
            [have_unit]
                side={sideNumber}
                role={roleName}
                search_recall_list=yes
            [/have_unit]
        [/not]
        [and]
            [have_unit]
                side={sideNumber}
                {unitFilter}
                search_recall_list=yes
            [/have_unit]
        [/and]
        [then]
            [role]
                side={sideNumber}
                role={roleName}
                {unitFilter}
            [/role]
        [/then]
    [/if]
```

GregoryLundberg commented:
How is that simplified? Is it faster?

Does an AND relationship in this case use short-circuit evaluation? That is, does the [AND] in your example avoid a second search of the recall list? The [ELSEIF] plainly does. I presume a search of the recall list is fairly costly and avoiding it should save quite a number of CPU cycles.

I presume that the interpreter is fairly straight forward and a tag equates to an op-code (sic). In that case, does the [ELSEIF] take significantly more space that [NOT] and [AND]? Or does it parse and evaluate every time it's executed and the more lines and tags of your example parse faster than the [ELSEIF] of mine.

I am still relatively new to understanding WML and appreciate all guidance.

Thank you.

CelticMinstrel commented:
I'm not entirely sure if `[and]` short-circuits, but one way in which Vultraz's version is simpler is the indentation level.

By the way, `[elseif]` isn't a shorthand for `[not][and]` even if it seems to have similar semantics in some cases. Also, if WML were a regular programming language like C++, it would be considered bad practice to have an if statement without the equivalent of the `[then]` tag. I dunno if this should apply to WML though.

GregoryLundberg commented:
Ye Gods I hope not. I started doing it because I saw it several places in mainline and core.

I would consider changing [and] and [or] to short-circuit a breaking change. It would be best to add new syntax.

Vultraz commented:
I try to fix any case of [then] omittance whenever it appears, so I'd prefer if you didn't introduce new cases.

GregoryLundberg commented:
OK. I will go through all my stuff and fix it where I did it.

CelticMinstrel commented:
> Every .. and I mean **EVERY** if-statement has **BOTH** a then-clause and an else-clause. Just because you don't see one doesn't mean it's not there.

Well no, the else-clause may truly not exist. At least in C-style programming languages (which includes C, C++, C#, Java, and JavaScript).

> I would consider changing [and] and [or] to short-circuit a breaking change. It would be best to add new syntax.

Huh? Why would it be a breaking change? As far as I know, there are no conditions that produce side-effects, so changing them to short-circuit should not alter the outcome in any way.

> OH ... and searching the recall list is probably linear but the [role]type= is documented to select the earlier listed type in preference to the following. A sophomoric implementation on 12 types in the clause would lead to 12 searches.

This is true. I think you probably can't get much better than this, though not bailing on first match like you described might improve the search a little. I'm not sure it's possible with the way filters currently work, though.

GregoryLundberg commented:
The point about it-then-else is not the language. It's the logic. On your flow-chart you drew a line with no box for one side. That line means 'omit this side of the clause' but how sure can one be that that is the right thing to do?

"If color is chosen then ..." vs "if color is chosen then ... else set default color" vs "if color is chosen then ... else throw" how sure are you that doing nothing at all is the right choice? Especially with newer programmers, you'd be surprised how many bugs they prevent when forced to actually think about it.

GregoryLundberg commented:
Can Lua-defined tags be used in conditionals? If not, why not? If so, short-circuit is a breaking change. Not that it should not happen, but there would be a risk.

Vultraz commented:
One can create custom conditional tags by registering a function in the `wesnoth.wml_conditionals` table. [More info.](https://wiki.wesnoth.org/LuaWML/Events#wesnoth.wml_conditionals)

GregoryLundberg commented:
To be clear, so I understand

    [if]
        [recall]id=Delfador[/recall]
            [then]
            ....

Will (a) never recall Delfador, and (b) always execute the [then] clause? Hmm .. I'd have thrown an exception, there, to kick the programmer's shins. Does it at least generate a warning on stderr?

Vultraz commented:
Pretty sure that's how it works yes, and with no error in stderr.

GregoryLundberg commented:
Just finished going through all my PR and fixing the [if] logic per your, did some DRY refactoring, fixed a couple bugs (both existing and in my changes), etc. I think I remembered all the comments and changed accordingly.

CelticMinstrel commented:
> Can Lua-defined tags be used in conditionals? If not, why not? If so, short-circuit is a breaking change. Not that it should not happen, but there would be a risk.

Lua-defined tags can be used, but that's new in 1.13 (as I recall), so I don't feel that changing it would be an issue.

GregoryLundberg commented:
With the recent changes to [role] I have a much cleaner branch for Liberty.

I am closing this PR in favor of my new PR #687.
